pkg/model/components/context.go

```diff
@@ -142,7 +153,59 @@ func Image(component string, clusterSpec *kops.ClusterSpec) (string, error) {
 	tag := strings.TrimSpace(string(b))
 	glog.V(2).Infof("Found tag %q for %q", tag, component)
 
-	return "gcr.io/google_containers/" + component + ":" + tag, nil
+	c := component + ":" + tag
+	return GetGoogleImageRepositoryContainer(clusterSpec, c), nil
+}
+
+func GetGoogleImageRepositoryContainer(clusterSpec *kops.ClusterSpec, c string) string {
+
+	// TODO validate container url
+
+	c = strings.TrimPrefix(c, "/")
+
+	if clusterSpec.AssetSpec != nil && clusterSpec.AssetSpec.DockerRepository == nil {
+		repo := *clusterSpec.AssetSpec.DockerRepository
+		return removeSlash(repo) + "/" + c
+	}
+
+	return GCR_IO + "/" + c
+}
+
+func GetGoogleFileRepositoryURL(clusterSpec *kops.ClusterSpec, u string) (string, error) {
+	u = strings.TrimPrefix(u, "/")
+
+	googleUrl := ""
+
+	if clusterSpec.AssetSpec != nil && clusterSpec.AssetSpec.FileRepository == nil {
+		googleUrl = removeSlash(*clusterSpec.AssetSpec.FileRepository) + "/" + u
+	}
+
+	if googleUrl == "" {
+		googleUrl = GCR_STORAGE + "/" + u
+	}
+
+	err := validateURL(googleUrl)
+
+	if err != nil {
+		return "", err
+	}
+
+	return googleUrl, err
+
+}
+
+func removeSlash(s string) string {
+	return strings.TrimSuffix(s, "/")
+}
+
+func validateURL(u string) error {
+	_, err := url.ParseRequestURI(u)
+
+	if err != nil {
+		glog.Warning("url is invalid %q", u)
+	}
+
+	return err
```

Comment:
Nit: `return fmt.Errorf("url is invalid %q", u)` would guarantee a nice error message here